# Proyecto SQL / SQL Project
El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

The coronavirus took the entire world by surprise, changing everyone's daily routines. City residents no longer spent their free time outside, going to cafés and shopping malls. Instead, more people stayed at home, reading books. This caught the attention of startups, which rushed to develop new applications for book lovers.

Pasos a seguir para desarrollar el proyecto / Steps to develop the project:
1. Describir los objetivos del estudio / Describe the study objectives.
2. Estudiar las tablas / Study the tables.
3. Realizar consultas SQL / SQL queries.
4. Conclusiones / Conclusions.

## Objetivo del estudio / Study Objective
Analizar la información dada para generar una propuesta de valor para un nuevo producto / Analyze the given information to generate a value proposition for a new product.

## Estudiar las tablas / Study the Tables

In [1]:
# Importar librerías
# Import libraries
import pandas as pd
from sqlalchemy import create_engine


# Configuración de conexión
# Connection Configuration
db_config = {'user': 'practicum_student', # Nombre de usuario / Username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # Contraseña / Password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # Puerto de conexión / Connection port
             'db': 'data-analyst-final-project-db'} # Nombre de la base de datos / Database name

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### `Books`

In [3]:
# Imprimir las primeras 5 filas de la tabla `books`
# Print the first 5 rows of the 'books' table
query = """
SELECT 
    *
FROM 
    books 
LIMIT 
    5;
"""

books = pd.io.sql.read_sql(query, con = engine)
books

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### `Authors`

In [4]:
# Imprimir las primeras 5 filas de la tabla `authors`
# Print the first 5 rows of the 'authors' table
query = """
SELECT 
    *
FROM 
    authors
LIMIT 
    5;
"""

authors = pd.io.sql.read_sql(query, con = engine)
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### `Publishers`

In [5]:
# Imprimir las primeras 5 filas de la tabla `publishers`
# Print the first 5 rows of the 'publishers' table
query = """
SELECT 
    *
FROM 
    publishers
LIMIT 
    5;
"""

publishers = pd.io.sql.read_sql(query, con = engine)
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### `Ratings`

In [6]:
# Imprimir las primeras 5 filas de la tabla `ratings`
# Print the first 5 rows of the 'ratings' table
query = """
SELECT 
    *
FROM 
    ratings
LIMIT 
    5;
"""

ratings = pd.io.sql.read_sql(query, con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### `Reviews`

In [7]:
# Imprimir las primeras 5 filas de la tabla `reviews`
# Print the first 5 rows of the 'reviews' table
query = """
SELECT 
    *
FROM 
    reviews
LIMIT 
    5;
"""

reviews = pd.io.sql.read_sql(query, con = engine)
reviews

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Consultas SQL / SQL queries

### Encontrar el número de libros publicados después del 1 de enero de 2000 / Find the number of books published after January 1, 2000

In [8]:
# Contar el número de libros publicados desde el 1 de enero de 2000
# Count the number of books published since January 1, 2000
query = """
SELECT 
    COUNT (*) as num_books
FROM 
    books
WHERE 
    publication_date > '2000-01-01';
"""

books_after_2000 = pd.io.sql.read_sql(query, con = engine)
books_after_2000

num_books
0        819

### Encontrar el número de reseñas de usuarios y la calificación promedio para cada libro / Find the number of user reviews and the average rating for each book

In [9]:
# Unir la tablas `books`, `reviews` y `ratings` 
# Join the 'books', 'reviews', and 'ratings' tables 

# Contar el número de reseñas y sacar el promedio de las calificaciones por libro 
# Count the number of reviews and calculate the average rating per book
query = """
SELECT 
    books.title AS title,
    COUNT(reviews.text) AS num_review,
    ROUND(AVG (ratings.rating), 2) AS avg_rating
FROM
    books
JOIN reviews ON reviews.book_id = books.book_id
JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.title;
"""

reviews_and_ratings = pd.io.sql.read_sql(query, con = engine)
reviews_and_ratings

title  num_review  avg_rating
0                            The Count of Monte Cristo         115        4.22
1                              Count Zero (Sprawl  #2)           4        2.50
2    The Botany of Desire: A Plant's-Eye View of th...           4        3.50
3                                 The Poisonwood Bible         110        4.36
4                                 The Canterbury Tales          18        3.33
..                                                 ...         ...         ...
988                           Of Love and Other Demons           4        4.50
989  In the Heart of the Sea: The Tragedy of the Wh...           9        3.33
990               Welcome to Temptation (Dempseys  #1)           4        5.00
991                      World's End (The Sandman  #8)           4        4.50
992                                  Holes (Holes  #1)         155        3.97

[993 rows x 3 columns]

### Identificar la editorial que ha publicado el mayor número de libros / Identify the publisher that has published the most books
Se tomarán en cuenta libros con más de 50 páginas para excluir folletos y publicaciones similares del análisis / Only books with more than 50 pages will be considered to exclude pamphlets and similar publications from the analysis.

In [10]:
# Contar el número de libros publicados por editorial
# Count the number of books published by each publisher
query = """
SELECT 
    publishers.publisher,
    COUNT(books.title) AS num_books
FROM
    books
JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE
    books.num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY 
    num_books DESC
LIMIT 
    1;
"""

publisher = pd.io.sql.read_sql(query, con = engine)
publisher

publisher  num_books
0  Penguin Books         42

### Identificar al autor que tiene la más alta calificación promedio del libro / Identify the author with the highest average book rating
Solo se analizan los libros con al menos 50 calificaciones / Only books with at least 50 ratings will be considered.

In [11]:
query = """
SELECT
    authors.author,
    AVG(ratings.rating) AS avg_rating,
    COUNT(ratings.rating) AS num_rating
FROM
    books
JOIN 
    authors ON authors.author_id = books.author_id
JOIN 
    ratings ON ratings.book_id = books.book_id
GROUP BY
    authors.author
HAVING 
    COUNT(ratings.rating) >= 50
ORDER BY
    avg_rating DESC
LIMIT
    1;
"""

author = pd.io.sql.read_sql(query, con = engine)
author

author  avg_rating  num_rating
0  Diana Gabaldon         4.3          50

### Encontrar el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros / Find the average number of text reviews among users who rated more than 50 books

In [12]:
# Promedio de reseñas entre los usuarios que calificaron más de 50 libros
# Average number of reviews among users who rated more than 50 books
query = """
WITH UsersWithMoreThan50Ratings AS
    (SELECT
        username,
        COUNT(rating)
    FROM
        ratings
    GROUP BY
        username
    HAVING
        COUNT(rating) > 50)
SELECT 
    ROUND(AVG(num_reviews), 2) AS avg_reviews
FROM
    (SELECT 
        reviews.username,
        COUNT(reviews.text) AS num_reviews
    FROM
        reviews
    JOIN 
        UsersWithMoreThan50Ratings ON UsersWithMoreThan50Ratings.username = reviews.username
    GROUP BY
        reviews.username) 
    AS review_filtered;
"""

avg_review = pd.io.sql.read_sql(query, con = engine)
avg_review

avg_reviews
0        24.33

## Conclusiones / Conclusions

- Se publicaron un total de 819 libros después del 1 de enero de 2000 / A total of `819` books were published after January 1, 2000.
- La editorial `Penguin Books` ha publicado el mayor número de libros, con `42` libros, lo que refleja su fuerte presencia en el mercado / The publisher `Penguin Books` has published the highest number of books, with `42` books, reflecting its strong presence in the market.
- La autora `Diana Gabaldon` tiene la calificación promedio más alta de `4.3` en sus libros, lo que muestra que sus obras son altamente apreciadas por los lectores / The author `Diana Gabaldon` has the highest average book rating of `4.3`, showing her books are highly appreciated by readers.
- El número promedio de reseñas de texto proporcionadas por los usuarios que calificaron más de 50 libros es de `24.33`, lo que sugiere que estos suarios están participando activamente con reseñas / The average number of text reviews provided by users who rated more than 50 books is `24.33`, suggesting that these users are actively engaging in reviews.